exercises/practice/circular-buffer/test/circular_buffer_test.exs

```diff
@@ -0,0 +1,237 @@
+defmodule CircularBufferTest do
+  use ExUnit.Case
+
+  def test_operation(buffer, %{"operation" => "read", "should_succeed" => ok} = op) do
+    read = CircularBuffer.read(buffer)
+
+    if ok do
+      assert {:ok, op["expected"]} == read
+    else
+      assert {:error, :empty} == read
+    end
+  end
+
+  def test_operation(buffer, %{"operation" => "write", "item" => item, "should_succeed" => ok}) do
+    write = CircularBuffer.write(buffer, item)
+
+    if ok do
+      assert :ok == write
+    else
+      assert {:error, :full} == write
+    end
+  end
+
+  def test_operation(buffer, %{"operation" => "overwrite", "item" => item}),
+    do: CircularBuffer.overwrite(buffer, item)
+
+  def test_operation(buffer, %{"operation" => "clear"}),
+    do: CircularBuffer.clear(buffer)
+
+  # @tag :pending
+  test "reading empty buffer should fail" do
+    capacity = 1
+    {:ok, buffer} = CircularBuffer.new(capacity)
+    operations = [%{"operation" => "read", "should_succeed" => false}]
+    Enum.each(operations, &test_operation(buffer, &1))
+  end
+
+  @tag :pending
+  test "can read an item just written" do
+    capacity = 1
+    {:ok, buffer} = CircularBuffer.new(capacity)
+
+    operations = [
+      %{"item" => 1, "operation" => "write", "should_succeed" => true},
+      %{"expected" => 1, "operation" => "read", "should_succeed" => true}
+    ]
+
+    Enum.each(operations, &test_operation(buffer, &1))
```

Comment:
I think it might be better for students if the whole test suite was a little bit less DRY and a bit more explicit. Something like this, so you don't need to go back to the top of the file to see the expected result:
```suggestion
    assert CircularBuffer.write(buffer, 1) == :ok
    assert CircularBuffer.read(buffer) == {:ok, 1}
```

Especially for beginners, seeing this list of maps with operations might be confusing.

Otherwise, in a non-teaching context, this would be all fine by me :)